sympy/solvers/solveset.py

```diff
@@ -999,6 +1007,281 @@ def _solveset(f, symbol, domain, _check=False):
+    modterms = list(f.atoms(Mod))
+
+    return (len(modterms) == 1 and  # only one Mod should be present
+            modterms[0].args[0].has(symbol) and  # B-> function of symbol
+            modterms[0].args[1].is_integer and  # C-> to be an integer.
+            any(isinstance(term, Mod)
+            for term in list(_term_factors(f)))  # free from other funcs
+            )
+
+
+def _invert_modular(modterm, rhs, n, symbol):
+    """
+    Helper function to invert modular equation.
+    ``Mod(a, m) - rhs = 0``
+
+    Generally it is inverted as (a, ImageSet(Lambda(n, m*n + rhs), S.Integers)).
+    More simplified form will be returned if possible.
+
+    If it is not invertible then (modterm, rhs) is returned.
+
+    Following cases arise:
+    1. a is x -> m*n + rhs is the required solution.
+    2. a is instance of Add -> we try to find two independent parts
+    of a and the symbol independent gets tranferred to other
+    side and again the solveset is called on the symbol dependent part.
+    3. a is instance of Mul -> same as we done in Add we separate out
+    the dependent and independent parts and tranfer the independent
+    part to rhs using invert here.
+    4. a is instance of Pow i.e a = base**exp -> 2 cases arise as following:
+    - a is (x_indep)**(x_dep) -> remainder is evaluated with the help
+    of discrete_log function and then the least period is being found out
+    with the help of totient function. For reference:
+    (https://en.wikipedia.org/wiki/Euler's_theorem)
+    - a is (x_dep)**(x_indep) -> we try to find all primitive solutions list
+    with the help of nthroot_mod function. m*n + rem is the general solution
+    where rem belongs to solutions list from nthroot_mod function.
+
+    Parameters
+    ==========
+
+    modterm, rhs : Expr
+        The modular equation to be inverted, ``modterm - rhs = 0``
+
+    symbol : Symbol
+        The variable in the equation to be inverted.
+
+    n : Dummy
+        Dummy variable for output g_n.
+
+    Returns
+    =======
+
+    A tuple (f_x, g_n) is being returned where f_x is modular independent function
+    of symbol and g_n being set of values f_x can have.
+
+    Examples
+    ========
+
+    >>> from sympy import symbols, exp, Mod, Dummy, S
+    >>> from sympy.solvers.solveset import _invert_modular as invert_modular
+    >>> x, y = symbols('x y')
+    >>> n = Dummy('n')
+    >>> invert_modular(Mod(exp(x), 7), S(5), n, x)
+    (Mod(exp(x), 7), 5)
+    >>> invert_modular(Mod(x, 7), S(5), n, x)
+    (x, ImageSet(Lambda(_n, 7*_n + 5), Integers))
+    >>> invert_modular(Mod(3*x + 8, 7), S(5), n, x)
+    (x, ImageSet(Lambda(_n, 7*_n + 6), Integers))
+    >>> invert_modular(Mod(x**4, 7), S(5), n, x)
+    (x, EmptySet())
+    >>> invert_modular(Mod(2**(x**2 + x + 1), 7), S(2), n, x)
+    (x**2 + x + 1, ImageSet(Lambda(_n, 3*_n + 1), Naturals0))
+
+    """
+    a, m = modterm.args
+
+    if not isinstance(a, (Dummy, Symbol, Add, Mul, Pow)):
```

Comment:
This does not seem right. You can go on to add a list of functions that are solvable (maybe in the near future). This should rather go at the last (L1190).

Reply:
I think it should go here because there is no point it should go at last because with this check there will not any need to check all the cases below it. Its like removing noise before entring the system so that user should be clear from the starting point that what cases he should consider in mind while reading the code below and not seeing this check at the end.

Reply:
> with this check there will not any need to check all the cases below it

This line is also checking all the cases. It's no difference checking here or below as `if's`.

> so that user should be clear from the starting point that what cases he should consider
 
It will be pretty much obvious to the user by just going through all the if cases: that what `a` could be.

Reply:
> This line is also checking all the cases. It's no difference checking here or below as if's

This line is meant to clear out all the cases which do not come below.
> It will be pretty much obvious to the user by just going through all the if cases

This is not a point that which thing is obvious, the point is to clear out the cases which are not currently handled by `_invert_modular`.

I am still not convinced that why this should go below. And moreover it is not causing any big problem that where this if should be placed. And in this case it will take short time(i know its very minute but still) than the if placed below.